### PR TITLE
osd: allow for injecting extra environment variables.

### PR DESCRIPTION
**Pull request by rzarzynski**
This patch is both stripped down (no support for extra argvs) and extended (envs at both prepare and main stages) version
of the Travis Nielsen's patch (https://github.com/rook/rook/pull/8763) to accommodate crimson-osd in Rook. Also, the naming convention isn't crimson-specific anymore as the mechanism could be useful with the classical OSD as well (e.g. tuning TCMalloc or even the dynamic linker).

The need for the patch comes from the fact that, although crimson finally exposes the same CLI interface as the classical
OSD, the broadly used development builds have ASan built in. As ASan, by default, complains if it isn't the very first loaded
DSO, we need to set the `ASAN_OPTIONS` environment variable to `verify_asan_link_order=0` to mitigate the early aborts.

Signed-off-by: Radoslaw Zarzynski <rzarzyns@redhat.com>

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

**Comment by rzarzynski** [comment]
When testing locally I saw a failure that **seems** unrelated:

```
$ make test
...
022-01-04 13:20:56.848289 I | sys: Output:
--- PASS: TestGetPartitions (0.00s)
=== RUN   TestParseUdevInfo
--- PASS: TestParseUdevInfo (0.00s)
=== RUN   TestListDevicesChildListDevicesChild
2022-01-04 13:20:56.848347 I | sys: command lsblk
--- PASS: TestListDevicesChildListDevicesChild (0.00s)
=== RUN   TestGrep
--- PASS: TestGrep (0.00s)
PASS
coverage: 34.4% of statements
ok      github.com/rook/rook/pkg/util/sys       (cached)        coverage: 34.4% of statements
?       github.com/rook/rook/pkg/version        [no test files]
FAIL
make[1]: *** [build/makelib/golang.mk:135: go.test.unit] Error 2
```

File: pkg/operator/ceph/cluster/osd/spec.go [hunk 1]
```diff
@@ -515,6 +516,13 @@ func (c *Cluster) makeDeployment(osdProps osdProperties, osd OSDInfo, provisionC
 		dataPath = activateOSDMountPath + osdID
 	}
 
+	// Inject extra envrinoment variables.
+	if extraEnvVarsValue, err := k8sutil.GetOperatorSetting(context.TODO(), c.context.Clientset, controller.OperatorSettingConfigMapName, "OSD_EXTRA_ENV_VARS", ""); err != nil {
```
Comment (leseb) [inline pkg/operator/ceph/cluster/osd/spec.go:520]:
```suggestion
	if extraEnvVarsValue, err := k8sutil.GetOperatorSetting(c.clusterInfo.Context, c.context.Clientset, controller.OperatorSettingConfigMapName, "OSD_EXTRA_ENV_VARS", ""); err != nil {
```

No need for a new context.

File: pkg/operator/ceph/cluster/osd/provision_spec.go [hunk 2]
```diff
@@ -230,6 +231,11 @@ func (c *Cluster) provisionOSDContainer(osdProps osdProperties, copyBinariesMoun
 	if osdProps.metadataDevice != "" {
 		envVars = append(envVars, metadataDeviceEnvVar(osdProps.metadataDevice))
 	}
+	if extraEnvVarsValue, err := k8sutil.GetOperatorSetting(context.TODO(), c.context.Clientset, controller.OperatorSettingConfigMapName, "OSD_EXTRA_ENV_VARS", ""); err != nil {
```
Comment (leseb) [inline pkg/operator/ceph/cluster/osd/provision_spec.go:234]:
```suggestion
	if extraEnvVarsValue, err := k8sutil.GetOperatorSetting(c.clusterInfo.Context, c.context.Clientset, controller.OperatorSettingConfigMapName, "OSD_EXTRA_ENV_VARS", ""); err != nil {
```

Comment (leseb) [inline pkg/operator/ceph/cluster/osd/provision_spec.go:234]: Why does the prepare job need this? I thought the args were needed only to start the OSD process.

File: pkg/operator/ceph/cluster/osd/spec.go [hunk 3]
```diff
@@ -515,6 +516,13 @@ func (c *Cluster) makeDeployment(osdProps osdProperties, osd OSDInfo, provisionC
 		dataPath = activateOSDMountPath + osdID
 	}
 
+	// Inject extra envrinoment variables.
```
Comment (leseb) [inline pkg/operator/ceph/cluster/osd/spec.go:519]:
```suggestion
	// Inject extra environment variables.
```

**Review by leseb — CHANGES_REQUESTED** [review]
Overall I'm not sure if using the operator config is the right approach since this will affect all the clusters deployed by the operator. Can't we detect that the OSD is a crimson OSD and then append the needed env variable? Just like we append `TCMALLOC_MAX_TOTAL_THREAD_CACHE_BYTES`?

File: pkg/operator/ceph/cluster/osd/spec.go (hunk 3 above)
Comment (rzarzynski) [reply pkg/operator/ceph/cluster/osd/spec.go:519]: Oops, thanks for pointing this out.

File: pkg/operator/ceph/cluster/osd/provision_spec.go (hunk 2 above)
Comment (leseb) [reply pkg/operator/ceph/cluster/osd/provision_spec.go:234]: Correct, so the `--mkfs` also needs this? If so, let's add a small comment in the code why the prepare stage needs it too :)

**Comment by rzarzynski** [comment]
> Overall I'm not sure if using the operator config is the right approach since this will affect all the clusters deployed by the operator. Can't we detect that the OSD is a crimson OSD and then append the needed env variable? Just like we append `TCMALLOC_MAX_TOTAL_THREAD_CACHE_BYTES`?

Thanks for the review!

I think it's beneficial the have the flexibility the manual approach brings (also for crimson-unrelated poking during development). Additionally, the auto-detection would be rather complex as it needs to not only differentiate between the classical OSD and crimson but also between crimson-dev and crimson-release builds. The target audience of this feature are mostly developers.

**Comment by leseb** [comment]
> > Overall I'm not sure if using the operator config is the right approach since this will affect all the clusters deployed by the operator. Can't we detect that the OSD is a crimson OSD and then append the needed env variable? Just like we append `TCMALLOC_MAX_TOTAL_THREAD_CACHE_BYTES`?
> 
> Thanks for the review!
> 
> I think it's beneficial the have the flexibility the manual approach brings (also for crimson-unrelated poking during development). Additionally, the auto-detection would be rather complex as it needs to not only differentiate between the classical OSD and crimson but also between crimson-dev and crimson-release builds. The target audience of this feature are mostly developers.

Understood

> > Overall I'm not sure if using the operator config is the right approach since this will affect all the clusters deployed by the operator. Can't we detect that the OSD is a crimson OSD and then append the needed env variable? Just like we append `TCMALLOC_MAX_TOTAL_THREAD_CACHE_BYTES`?
> 
> Thanks for the review!
> 
> I think it's beneficial the have the flexibility the manual approach brings (also for crimson-unrelated poking during development). Additionally, the auto-detection would be rather complex as it needs to not only differentiate between the classical OSD and crimson but also between crimson-dev and crimson-release builds. The target audience of this feature are mostly developers.

If it's a dev flow only, how about we go with the following approach:

* On every container definition use `EnvFrom`
* Where `EnvFrom` points to a CM using `ConfigMapRef` with a `Name` and mark `Optional: true` in case the CM does not exist

We can call this configmap `rook-ceph-daemon-env-override`?
What do you think?

**Comment by rzarzynski** [comment]
After the very first reading the idea looks promising! I will be experimenting with it to see whether this could a matter of just a dedicated file for crimson in the `examples` directory.

**Comment by rzarzynski** [comment]
Closing as #9647 got merged.

**Pull request closed without merging**